parsl/executors/high_throughput/mpi_resource_management.py

```diff
@@ -0,0 +1,191 @@
+        self.requested = requested
+        self.available = available
+
+    def __str__(self):
+        return f"MPIResourceUnavailable(requested={self.requested} available={self.available}"
+
+
+class NoopScheduler:
+    def __init__(
+        self,
+        pending_task_q: multiprocessing.Queue,
+        pending_result_q: multiprocessing.Queue,
+    ):
+        self.pending_task_q = pending_task_q
+        self.pending_result_q = pending_result_q
+
+    def put_task(self, task) -> None:
+        return self.pending_task_q.put(task)
+
+    def get_result(self, block: bool, timeout: float):
+        return self.pending_result_q.get(block, timeout=timeout)
+
+    def put_result(self, result) -> None:
+        return self.pending_result_q.put(result)
+
+
+class MPITaskScheduler(NoopScheduler):
+    def __init__(
+        self,
+        pending_task_q: multiprocessing.Queue,
+        pending_result_q: multiprocessing.Queue,
+    ):
+        super().__init__(pending_task_q, pending_result_q)
+        self.scheduler = identify_scheduler()
+        # PriorityQueue is threadsafe
+        self._backlog_queue: queue.PriorityQueue = queue.PriorityQueue()
+        self._map_tasks_to_nodes: Dict[str, List[str]] = {}
+        self.available_nodes = get_nodes_in_batchjob(self.scheduler)
+        self._free_node_counter = SpawnContext.Value("i", len(self.available_nodes))
+        # mp.Value has issues with mypy
+        # issue https://github.com/python/typeshed/issues/8799
+        # from mypy 0.981 onwards
+        self.nodes_q = SpawnContext.Queue()
+        for node in self.available_nodes:
+            self.nodes_q.put(node)
+
+        logger.info(
+            f"Starting MPITaskScheduler with {len(self.available_nodes)}"
+        )
+
+    def _get_nodes(self, num_nodes: int):
+        logger.debug(
+            f"Requesting : {num_nodes=} we have {self._free_node_counter}"
+        )
+        acquired_nodes = []
+        with self._free_node_counter.get_lock():
+            if num_nodes <= self._free_node_counter.value:  # type: ignore[attr-defined]
+                self._free_node_counter.value -= num_nodes  # type: ignore[attr-defined]
+            else:
+                raise MPIResourceUnavailable(
+                    requested=num_nodes, available=self._free_node_counter.value  # type: ignore[attr-defined]
+                )
+
+            for i in range(num_nodes):
+                node = self.nodes_q.get()
+                acquired_nodes.append(node)
+        return acquired_nodes
+
+    def _return_nodes(self, nodes: List[str]) -> None:
+        for node in nodes:
+            self.nodes_q.put(node)
+        with self._free_node_counter.get_lock():
+            self._free_node_counter.value += len(nodes)  # type: ignore[attr-defined]
+
+    def put_task(self, task_package: dict):
+        user_ns = locals()
+        user_ns.update({"__builtins__": __builtins__})
+        _f, _args, _kwargs, resource_spec = unpack_res_spec_apply_message(
```

Comment:
I talked with @yadudoc elsewhere a bit about having the resource spec sent separately to the apply message, to avoid deserialization of the potentially heavy apply message here inside the manager code: instead extract it in the `HighThroughputExecutor` in the submitting process, and send it alongside all the task messages.

That could avoid deserializing the function and arguments inside the manager, which I think is quite a serious performance concern.